api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/authenticate/ForgotPassword.java

```diff
@@ -0,0 +1,321 @@
+package edu.cornell.mannlib.vitro.webapp.controller.authenticate;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+
+import javax.mail.Message;
+import javax.servlet.ServletContext;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServletRequest;
+
+import edu.cornell.mannlib.vitro.webapp.beans.ApplicationBean;
+import edu.cornell.mannlib.vitro.webapp.beans.UserAccount;
+import edu.cornell.mannlib.vitro.webapp.config.ConfigurationProperties;
+import edu.cornell.mannlib.vitro.webapp.controller.VitroRequest;
+import edu.cornell.mannlib.vitro.webapp.controller.freemarker.FreemarkerHttpServlet;
+import edu.cornell.mannlib.vitro.webapp.controller.freemarker.UrlBuilder;
+import edu.cornell.mannlib.vitro.webapp.controller.freemarker.responsevalues.ResponseValues;
+import edu.cornell.mannlib.vitro.webapp.controller.freemarker.responsevalues.TemplateResponseValues;
+import edu.cornell.mannlib.vitro.webapp.dao.UserAccountsDao;
+import edu.cornell.mannlib.vitro.webapp.dao.WebappDaoFactory;
+import edu.cornell.mannlib.vitro.webapp.email.FreemarkerEmailFactory;
+import edu.cornell.mannlib.vitro.webapp.email.FreemarkerEmailMessage;
+import edu.cornell.mannlib.vitro.webapp.i18n.I18n;
+import edu.cornell.mannlib.vitro.webapp.i18n.I18nBundle;
+import edu.cornell.mannlib.vitro.webapp.modelaccess.ModelAccess;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+@WebServlet(name = "forgot-password", urlPatterns = {"/forgot-password"})
+public class ForgotPassword extends FreemarkerHttpServlet {
+
+    private static final String RESET_PASSWORD_URL = "/accounts/resetPassword";
+
+    private static final int DAYS_TO_USE_PASSWORD_LINK = 5;
+
+    private static final String TEMPLATE_NAME = "userAccounts-resetPasswordRequest.ftl";
+
+    private static final Log log = LogFactory.getLog(ForgotPassword.class.getName());
+
+
+    @Override
+    protected ResponseValues processRequest(VitroRequest vreq) throws Exception {
+        // Random time interval sleep so attacker can't calculate whether provided email is bound or not
+        sleepForRandomTime();
+
+        Map<String, Object> dataContext = new HashMap<>();
+        setCommonValues(dataContext, vreq);
+        UserAccountsDao userAccountsDao = constructUserAccountsDao(vreq);
+        I18nBundle i18n = I18n.bundle(vreq);
+
+        boolean isEnabled = isFunctionalityEnabled(vreq);
+        dataContext.put("isEnabled", isEnabled);
+        if (!isEnabled || vreq.getMethod().equalsIgnoreCase("GET")) {
+            return showForm(dataContext);
+        }
+
+        String captchaInput = vreq.getParameter("defaultReal");
+        String captchaDisplay = vreq.getParameter("defaultRealHash");
+        if (!captchaHash(captchaInput).equals(captchaDisplay)) {
```

Comment:
So if the form received a pair of defaultReal value and defaultRealHash created as in the captchaHash method, then password reset will be allowed? If so, this check can be bypassed very easily.

Reply:
I agree but, unfortunately, this is the only way it can be done with this implementation. We can maybe open a separate issue for captcha rework or to include 3rd party captcha service. I will be happy if given an opportunity to work on it.

Reply:
probably this will be a separated issue which will fix captcha here and at the feedback form.